libraries/chain/include/eosio/chain/finality/finalizer.hpp

```diff
@@ -71,41 +71,44 @@ namespace eosio::chain {
       using fsi_map = std::map<bls_public_key, fsi_t>;
 
    private:
-      const block_timestamp_type        t_startup;             // nodeos startup time, used for default safety_information
       const std::filesystem::path       persist_file_path;     // where we save the safety data
       mutable std::mutex                mtx;
       mutable fc::datastream<fc::cfile> persist_file;          // we want to keep the file open for speed
       std::map<bls_public_key, finalizer>  finalizers;         // the active finalizers for this node, loaded at startup, not mutated afterwards
       fsi_map                           inactive_safety_info;  // loaded at startup, not mutated afterwards
       fsi_t                             default_fsi = fsi_t::unset_fsi(); // default provided at spring startup
       mutable bool                      inactive_safety_info_written{false};
+      bool                              enable_voting = false;
 
    public:
-      my_finalizers_t(block_timestamp_type startup_time, const std::filesystem::path& persist_file_path)
-         : t_startup(startup_time)
-         , persist_file_path(persist_file_path)
+      explicit my_finalizers_t(const std::filesystem::path& persist_file_path)
+         : persist_file_path(persist_file_path)
       {}
 
       template<class F> // thread safe
-      void maybe_vote(const finalizer_policy& fin_pol,
-                      const block_state_ptr& bsp,
-                      const digest_type& digest,
-                      F&& process_vote) {
+      void maybe_vote(const block_state_ptr& bsp, F&& process_vote) {
 
          if (finalizers.empty())
             return;
 
+         assert(bsp->active_finalizer_policy);
+         const auto& fin_pol = *bsp->active_finalizer_policy;
+
          std::vector<vote_message_ptr> votes;
          votes.reserve(finalizers.size());
 
          // Possible improvement in the future, look at locking only individual finalizers and releasing the lock for writing the file.
          // Would require making sure that only the latest is ever written to the file and that the file access was protected separately.
          std::unique_lock g(mtx);
 
+         if (!enable_voting && bsp->timestamp() < fc::time_point::now() - fc::seconds(30))
+            return;
+         enable_voting = true;
```

Comment:
Maybe you can re-order the condition above such that `enable_voting = true;` is not always executed once it is true.